[Azure Maps] Rendering API review

dubiety opened the pull request:
# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

[APIView for Azure.Maps.Render](https://apiview.dev/Assemblies/Review/ec51a1ccfeed4e9ca513f34388275659)

dubiety commented:
/azp run net - maps - tests

dubiety commented:
/azp run net - AzureMapsRender- tests

dubiety commented:
/azp run net - AzureMapsRender- ci

dubiety commented:
/azp run prepare-pipelines

dubiety commented:
/azp run net - maps - ci

dubiety commented:
/azp run net - maps - test

dubiety commented:
/azp run net - maps - ci

dubiety commented:
/azp run net - maps - test

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[Azure.Maps.Render](https://apiview.dev/Assemblies/Review/e2c2b4239085400bbec0e35d10752270)

dubiety commented:
/azp run net - maps - ci

dubiety commented:
/azp run net - maps - test

dubiety commented:
/azp run net - maps - ci

dubiety commented:
/azp run net - maps - tests

dubiety commented:
/azp run net - maps - tests

dubiety commented:
/azp run net - maps - tests

dubiety commented:
/azp run net - maps - tests

dubiety commented:
/azp run net - maps - tests

dubiety committed:
Update API file

dubiety commented:
/azp run net - maps - tests

dubiety commented:
/azp run net - maps - tests

dubiety commented:
/azp run net - maps - tests

dubiety commented:
/azp run net - Azure.Maps.Render - ci

dubiety commented:
/azp run net - maps - ci

dubiety commented:
/azp run net - maps - tests

dubiety commented:
/check-enforcer evaluate

dubiety commented:
/check-enforcer override

dubiety commented:
/azp run prepare-pipelines

dubiety commented:
/azp run net - maps - tests

dubiety commented:
/azp run net - maps - ci

dubiety commented:
/azp run net - maps - tests

dubiety commented:
/azp run net - maps - tests

dubiety commented:
/azp run net - maps - tests

dubiety commented:
/azp run net - maps - tests

dubiety committed:
Add PushpinStyle and PathStyle

benbp commented on sdk/maps/test-resources.json:
```diff
@@ -0,0 +1,105 @@
+            "type": "string",
+            "value": "[resourceGroup().Name]"
+        },
+        "TENANT_ID": {
+            "type": "string",
+            "value": "[parameters('tenantId')]"
+        },
+        "CLIENT_ID": {
+            "type": "string",
+            "value": "[parameters('testApplicationId')]"
+        },
+        "CLIENT_SECRET": {
+            "type": "string",
+            "value": "[parameters('testApplicationSecret')]"
+        },
```
These should not be exported and are already available via `MAPS_CLIENT_SECRET`, etc.

Commit:
Apply best practice

Commit:
Add Azure.Maps.Common project and fix all review comments

Commit:
Update test cases

jsquire commented on sdk/maps/Azure.Maps.Render/src/Models/Options/RenderStaticImageOptions.cs:
```diff
@@ -0,0 +1,91 @@
+        /// <summary> Desired zoom level of the map. Zoom value must be in the range: 0-20 (inclusive). Default value is 12. Please see <see href="https://docs.microsoft.com/azure/location-based-services/zoom-levels-and-tile-grid">Zoom Levels and Tile Grid</see> for details. </summary>
+        public int? ZoomLevel { get; set; }
+        /// <summary>
+        /// Coordinates of the center point. Format: 'lon,lat'. Projection used EPSG:3857.
```
This all seems very informative but also assumes a good deal of base knowledge.   I don't know that this content is going to be accessible or helpful to developers looking to onboard to the library.

jsquire commented on sdk/maps/Azure.Maps.Render/src/Models/PushpinStyle.cs:
```diff
@@ -0,0 +1,170 @@
+
+            if (LabelAnchorShiftInPixels != null)
+            {
+                sb.Append($"|la{LabelAnchorShiftInPixels?.X} {LabelAnchorShiftInPixels?.Y}");
```
I beleive using a bling string here is going to cause unnecessary allocations.   I'd suggest either doing these as individual `Append` calls or using `AppendFormat`.   Likewise for the below.

jsquire commented on sdk/maps/Azure.Maps.Render/src/properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,13 @@
+
+// Replace with test project/test project public key and uncomment to make internal members visible to
+// your test project. If not needed, this can be deleted.
+// [assembly: InternalsVisibleTo("Azure.Template.Generated.Tests, PublicKey=0024000004800000940000000602000000240000525341310004000001000100d15ddcb29688295338af4b7686603fe614abd555e09efba8fb88ee09e1f7b1ccaeed2e8f823fa9eef3fdd60217fc012ea67d2479751a0b8c087a4185541b851bd8b16f8d91b840e51b1cb0ba6fe647997e57429265e85ef62d565db50a69ae1647d54d7bd855e4db3d8a91510e5bcbd0edfbbecaa20a7bd9ae74593daa7b11b4")]
```
If this isn't needed, we should consider removing it.

jsquire commented on sdk/maps/Azure.Maps.Render/tests/Samples/RenderClientSamples.cs:
```diff
@@ -0,0 +1,208 @@
+            #region Snippet:RenderImageryTiles
+            var credential = new DefaultAzureCredential();
+            var clientId = TestEnvironment.MapAccountClientId;
+            var client = new MapsRenderClient(credential, clientId);
```
The same pattern is recommended for other snippets to avoid referencing test constructs in samples.

jsquire commented on sdk/maps/Azure.Maps.Render/tests/Samples/RenderClientSamples.cs:
```diff
@@ -0,0 +1,208 @@
+                Assert.IsTrue(fileStream.Length > 0);
+            }
+            #endregion
+            #endregion
```
```suggestion
```

jsquire replied:
I think this is a duplicate?   Maybe I'm misreading?

dubiety replied:
It's a embedded `#region` here. I believe it's correct here :)

dubiety committed:
Try to upload a fake README and test pipeline works or not

dubiety commented:
/azp run net - maps - ci

dubiety commented:
/azp run net - maps - ci

dubiety commented:
/azp run net - maps - tests

dubiety commented:
/azp run net - maps - tests

dubiety commented:
/azp run net - maps - tests

dubiety commented:
/azp run net - maps - tests

dubiety commented:
/azp run net - maps - tests

dubiety commented:
/azp run net - maps/Azure.Maps.Render - tests

dubiety commented:
/azp run net - Azure.Maps.Render - tests

dubiety commented:
/azp run net - Azure.Maps.Rendering - tests

dubiety commented:
/azp run net - maps - tests

dubiety commented:
/azp run net - maps - tests

jsquire commented on sdk/maps/Azure.Maps.Common/src/LocalizedMapView.cs:
```diff
@@ -40,37 +40,37 @@ public LocalizedMapView(string value)
+        public static LocalizedMapView Pakistan { get; } = new LocalizedMapView(PKValue);
         /// <summary> Palestinian Authority (Arabic View). </summary>
-        public static LocalizedMapView PS { get; } = new LocalizedMapView(PSValue);
+        public static LocalizedMapView Palestinian { get; } = new LocalizedMapView(PSValue);
```
Should this be `Palestine`?

dubiety replied:
Should be `PalestinianAuthority`. Thanks for remidning!

jsquire commented on sdk/maps/Azure.Maps.Rendering/samples/MapsRenderTilesImageriesImagesSamples.md:
```diff
@@ -0,0 +1,186 @@
+
+The imagery will look like:
+
+![BerlinImagery](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/maps/Azure.Maps.Rendering/tests/BerlinImagery.png?raw=true)
```
nit:  We should have ALT text in these to support screen readers.

jsquire commented on sdk/maps/Azure.Maps.Rendering/src/Models/Options/GetMapStaticImageOptions.cs:
```diff
@@ -0,0 +1,91 @@
+        /// </summary>
+        public Azure.Maps.LocalizedMapView? LocalizedMapView { get; set; }
+
+#pragma warning disable CA2227 // Collection properties should be read only
```
@KrzysztofCwalina, @tg-msft:   Would you please offer your thoughts on the settable collection properties?

dubiety replied:
To put more contents here. We make the settable collection because the lists are optional parameters. We don't want user to set the lists from ctor or create an empty list even when they are not used.

jsquire replied:
I've approved the PR, but please resolve this with the architects.   Settable collection properties are contrary to guidelines, which is why there's an analyzer rule to enforce this constraint.

dubiety replied:
I've removed pragma warning and make it a read only list.

jsquire commented on sdk/maps/Azure.Maps.Rendering/src/Models/RegionalCopyright.cs:
```diff
@@ -0,0 +1,35 @@
+        }
+
+        /// <summary> Initializes a new instance of RegionalCopyright. </summary>
+        /// <param name="copyrights"> Copyrights array. </param>
```
nit:  The comments for the parameters lack enough context to be helpful.   The use of `array` here is also inaccurate given the type.

jsquire commented on sdk/maps/Azure.Maps.Rendering/tests/Samples/RenderClientSamples.cs:
```diff
@@ -0,0 +1,211 @@
+        [Test]
+        public void RenderingMapTiles()
+        {
+            TokenCredential credential = new DefaultAzureCredential();
```
These all should be `TestEnvironment.Credential` unless contained within an `#if SNIPPET` scope.

jsquire reviewed (commented):
Please remove `FriendAssemblies.publickey` from the `tests` directory; only the repository signing should be used.

dubiety commented:
/azp run net - maps - tests

jsquire commented on sdk/maps/Azure.Maps.Search/tests/ReverseSearchAddressTests.cs:
```diff
@@ -78,10 +79,10 @@ public async Task CanPoolReverseSearchAddressBatch()
                 new ReverseSearchAddressQuery(new ReverseSearchOptions { coordinates = new GeoPosition(-122.333345, 47.606038) }),
             });
-
+            Thread.Sleep(500);
```
This should be `Task.Delay` since you're running in an asynchronous context.   That said, if we can make the test approach more deterministic rather than resorting to arbitrary delays, that'd be a much better idea.

Why does it matter there is a delay between the LRO call and waiting for completion?   I'd expect customer code to do exactly what this test had been doing.

dubiety replied:
We always have the unit test error in MacOS environment for LRO. Like this one: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1832647&view=logs&j=e1233b80-fe26-54b1-8da9-2330cdd5d8f1&t=e42962a0-eaf5-52a0-a3b4-15bb9b250940&l=706
It's flacky, but I don't know why. If we sleep for few milliseconds, we should be able the fix this issue. Do you have any thoughts why LRO only fails in MacOS environment? Is there any way we can avoid or fix this issue if we don't add `Task.Delay`?

jsquire replied:
There's no known issue of which I'm aware.  Without context of some sort, there's nothing that I can even offer speculation for.  Do you have a stacktrace or other information around what you believe is different specific to macOS behavior?

dubiety replied:
I guess it's performance/timing issue. If we send LRO request and ask for result immediately (like what we have in unit test), and server side hasn't finish processing the request yet, it can result in 404 request not found. I've added `Task.Dealy`. Should avoid the flacky test result now.

dubiety commented:
/azp run net - maps - tests

dubiety commented:
/azp run net - maps - tests

dubiety commented:
> Please remove `FriendAssemblies.publickey` from the `tests` directory; only the repository signing should be used.

Removed

dubiety commented:
/azp run net - maps - tests

dubiety committed:
Update changelog

dubiety committed:
Fix comments and update LocalizedMapView

dubiety committed:
Add Thread.Sleep to avoid unit test failure

dubiety committed:
Render -> Rendering

Commit:
Update code by comments and update supporting version to 2022-08-01

dubiety committed:
Update API view and samples

dubiety commented on sdk/maps/Azure.Maps.Render/src/Models/Options/RenderStaticImageOptions.cs:
```diff
@@ -0,0 +1,91 @@
+        /// </summary>
+        public Azure.Maps.LocalizedMapView? LocalizedMapView { get; set; }
+
+#pragma warning disable CA2227 // Collection properties should be read only
```
I removed the pragma warning and make it a read only element.

dubiety committed:
Fix all review feedbacks

dubiety committed:
Update variable to avoid spelling error

dubiety commented:
/azp run net - maps - tests

dubiety committed:
Initial commit for Maps Render SDK

Add CI and test yaml files

Fix RenderClient compiling errors

Finished modification of RenderClient

Support Copyright test

Update changelog.md

Complete all render APIs

Update all render API

Update export API

Refine API version parameter

Add GetMapsTests unit test cases

Add more constructor for RenderClient

Update constructor signature

Fix copyright warning

Remove unused files

Change class name from Copyright to RenderCopyrights

Add ci.yml, tests.yml and test-reosurces.json for CI and pipeline

Update ci.yml, tests.yml and test-resources.json

Add shared source file in Azure.Core

Fix CI errors

Update README.md

Fix cspell issues

Fix CI errors

Try to add cspell.json file

Fix Analyze error

Fix CI errors, update samples

Fix CI errors

Update API interface file

Fix cspell errors

Fix CI errors - Wrong snippet name

Remove unused sample file

Fix CI errors

Update snippet values

Update code snippets

Update ci to use SdkType = client

Remove SdkType from ci.yml

Remove SDKType:client from tests.yml

Try to use subscription key to run live tests

Use optional variable to fetch SUBSCRIPTION_KEY

Use CLIENT_SECRET instead of SUBSCRIPTION_KEY

Use SUBSRIPTION_KEY from environment variable

Use client secret to run live test

Delete unused file: cspell.json

Update test records

dubiety committed:
Make test case more general

Commit:
Remove unused env vars in tests.yml

dubiety commented:
/azp run net - maps - tests

dubiety commented:
/azp run net - maps - tests

dubiety commented:
/azp run net - maps - tests

Commit:
Update README in maps root folder

dubiety commented:
/azp run net - maps - ci
